### DISCUSSION TITLE
Shaders

### DISCUSSION THREAD

**Discussion by TopperDEL** [StereoKit/StereoKit#112] [Q&A] [answered]
I have absolutely no knowledge about shaders. But if e.g. I wanted a sphere to "glow" or "bloom", this should be possible by adding a shader. If I won't or cannot code that on my own - how could I use another shader here? Do you know of any platform where I can search for shaders?
A quick google-search lead me to a site from NVIDIA where I could download "hlsl"-shaders - but they have a different file type and lead to an exception the way I tried to use it.

So in short: do you have a beginners way/guide of adding some "cool" shaders or is there a source to search for stuff that should work with StereoKit?

**Comment by maluoi** [comment, marked as answer]
Since v0.3 involves a complete overhaul of the graphics system, I haven't written up the docs for shaders quite yet! A lot is the same, but there's a few differences. It's real HLSL, but engine hooks always mean that porting a shader around is never a trivial task. Shaders generally have to be built per-engine.

I've got a number of [builtin shaders](https://github.com/maluoi/StereoKit/tree/master/StereoKitC/shaders_builtin) that can be used for reference, but the format will change in v0.3 to [code like this](https://github.com/maluoi/StereoKit/tree/develop/StereoKitC/shaders_builtin).

Another note here is that the bloom technique in particular is typically done as a full-screen / post-process effect. This is _really_ bad for performance on mobile systems like the HoloLens, and so StereoKit doesn't support anything like that yet. If I add the functionality, I would likely disable it on mobile to prevent people from shooting themselves in the foot.

Depending on what type of glow you're looking for, there's other, better ways to approach this! If it's an internal glow, this can generally be done easily with a shader as custom lighting. For external glow, it's common to use billboard quads with transparency, or shells with transparency. Additive transparency is usually best for this! v0.2 only has a more 'normal' blend, but additive is an easy add for v0.3, I'll actually tack that in right now. (There's a lot of mental complexity in alpha blending, so I haven't decided the best way to give all of it to users yet. The current way is -very- easy to understand, but not comprehensive.)

Also FYI, I did put up a [preview build](https://www.nuget.org/packages/StereoKit/0.3.0-preview.1) of v0.3 yesterday. There are definitely differences, and the docs won't be updated until I fully release, but it shouldn't be too tricky to figure some of it out. The [test project](https://github.com/maluoi/StereoKit/tree/develop/Examples/StereoKitTest/Demos) is also up-to-date, so it's good for reference!

**Reply by maluoi** [reply]
Additive transparency is in f6f805d03cff5b69204950b696229cf4194de80b now, so that'll be in the next v0.3 build! I'm not in a rush to push up another build right now, but feel free to poke me if you want it sooner rather than later :)

**Reply by TopperDEL** [reply]
Thank! I just need to explore some fields to get an understanding how things work. I still have plenty to do around making it visually appealung - so there is no immediate demand here.
But thank you so much for your support! I appreciate it!
